app/src/main/cpp/skyline/services/codec/IHardwareOpusDecoder.cpp

```diff
@@ -0,0 +1,126 @@
+        if (result != OPUS_OK)
+            throw OpusException(result);
+    }
+
+    IHardwareOpusDecoder::~IHardwareOpusDecoder() {
+        opus_decoder_destroy(decoderState);
+    }
+
+    Result IHardwareOpusDecoder::DecodeInterleavedOld(type::KSession &session, ipc::IpcRequest &request, ipc::IpcResponse &response) {
+        auto opusDataIn{request.inputBuf.at(0)};
+        auto pcmDataOut{request.outputBuf.at(0).cast<opus_int16>()};
+
+        auto decodeResult{Decode(opusDataIn, pcmDataOut)};
+        if (!decodeResult) {
+            throw OpusException(decodeResult.result);
+        }
+
+        response.Push(decodeResult->decodedDataSize);
+        response.Push(decodeResult->decodedSampleCount);
+        return {};
+    }
+
+    Result IHardwareOpusDecoder::DecodeInterleavedWithPerfOld(type::KSession &session, ipc::IpcRequest &request, ipc::IpcResponse &response) {
+        auto opusDataIn{request.inputBuf.at(0)};
+        auto pcmDataOut{request.outputBuf.at(0).cast<opus_int16>()};
+
+        auto decodeResult{Decode(opusDataIn, pcmDataOut)};
+        if (!decodeResult) {
+            throw OpusException(decodeResult.result);
+        }
+
+        response.Push(decodeResult->decodedDataSize);
+        response.Push(decodeResult->decodedSampleCount);
+        response.Push<u64>(0);
+        return {};
+    }
+
+    Result IHardwareOpusDecoder::DecodeInterleavedWithPerfAndResetOld(type::KSession &session, ipc::IpcRequest &request, ipc::IpcResponse &response) {
+        return DecodeInterleaved(session, request, response);
+    }
+
+    Result IHardwareOpusDecoder::DecodeInterleaved(type::KSession &session, ipc::IpcRequest &request, ipc::IpcResponse &response) {
+        auto reset{request.Pop<bool>()};
+        auto opusDataIn{request.inputBuf.at(0)};
+        auto pcmDataOut{request.outputBuf.at(0).cast<opus_int16>()};
+
+        if (reset)
+            ResetContext();
+
+        auto decodeResult{Decode(opusDataIn, pcmDataOut)};
+        if (!decodeResult) {
+            throw OpusException(decodeResult.result);
+        }
+
+        response.Push(decodeResult->decodedDataSize);
+        response.Push(decodeResult->decodedSampleCount);
+        response.Push<u64>(0);
+        return {};
+    }
+
+    void IHardwareOpusDecoder::ResetContext() {
+        opus_decoder_ctl(decoderState, OPUS_RESET_STATE);
+    }
+
+    ResultValue<IHardwareOpusDecoder::DecodeResult> IHardwareOpusDecoder::Decode(span<u8> dataIn, span<opus_int16> dataOut) {
+        if (dataIn.size() <= sizeof(OpusDataHeader)) {
+            return result::BadOpusDataHeader;
+        }
+
+        auto frameSize{util::SwapEndianness(dataIn.as<OpusDataHeader>().size)};
+        auto requiredOutSize{frameSize + sizeof(OpusDataHeader)};
+        if (frameSize > decoderOutputBufferSize || dataOut.size() < requiredOutSize) {
+            return result::OpusBufferTooSmall;
+        }
+
+        dataIn = dataIn.subspan(sizeof(OpusDataHeader));
```

Comment:
make this a new variable 'sampleDataIn' for something otherwise it's a tad confusing

Reply:
added a comment instead

Reply:
I think a new variable here is a better idea